src/layers/image/rotation.cpp

```diff
@@ -0,0 +1,119 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#define LBANN_ROTATION_LAYER_INSTANTIATE
+#include "lbann/layers/image/rotation.hpp"
+
+#include <math.h>  
+
+namespace lbann {
+
+template <typename TensorDataType, data_layout Layout, El::Device Device>
+void rotation_layer<TensorDataType, Layout, Device>::fp_compute() {
+
+  // Useful constants
+  constexpr DataType Pi = M_PI;
+  constexpr DataType degree = 180;
+  constexpr DataType half = 0.5;
+  constexpr DataType zero = 0;
+  
+  // Input and output tensors
+  const auto& local_input = this->get_local_prev_activations();
+  auto& local_output = this->get_local_activations();
+
+  // Tensor dimensions
+  const auto& input_dims = this->get_input_dims(0);
+  const El::Int num_dims = input_dims.size();
+  const auto& num_samples = local_input.Width();
+  const El::Int num_channels = std::accumulate(input_dims.begin(),
+                                               input_dims.end()-2,
+                                               1,
+                                               std::multiplies<int>());
+  const El::Int input_height = input_dims[num_dims-2];
+  const El::Int input_width = input_dims[num_dims-1];
```

Comment:
We can simplify if we assume the data is in CHW format:
```suggestion
  const El::Int num_channels = input_dims[0]
  const El::Int input_height = input_dims[1];
  const El::Int input_width = input_dims[2];
```